src/shared/creds-util.c

```diff
@@ -52,3 +69,843 @@ int read_credential(const char *name, void **ret, size_t *ret_size) {
+                else if (r > 0)
+                        log_debug("Running in container, not attempting to use TPM2.");
+
+                try_tpm2 = r <= 0;
+        }
+
+        if (try_tpm2 ||
+            sd_id128_in_set(with_key, CRED_AES256_GCM_BY_TPM2_HMAC, CRED_AES256_GCM_BY_HOST_AND_TPM2_HMAC)) {
+
+                r = tpm2_seal(tpm2_device,
+                              tpm2_pcr_mask,
+                              &tpm2_key,
+                              &tpm2_key_size,
+                              &tpm2_blob,
+                              &tpm2_blob_size,
+                              &tpm2_policy_hash,
+                              &tpm2_policy_hash_size);
+                if (r < 0) {
+                        if (!sd_id128_is_null(with_key))
+                                return r;
+
+                        log_debug_errno(r, "TPM2 sealing didn't work, not using: %m");
+                }
+
+                assert(tpm2_blob_size <= CREDENTIAL_FIELD_SIZE_MAX);
+                assert(tpm2_policy_hash_size <= CREDENTIAL_FIELD_SIZE_MAX);
+        }
+#endif
+
+        if (sd_id128_is_null(with_key)) {
+                /* Let's settle the key type in auto mode now. */
+
+                if (host_key && tpm2_key)
+                        id = CRED_AES256_GCM_BY_HOST_AND_TPM2_HMAC;
+                else if (tpm2_key)
+                        id = CRED_AES256_GCM_BY_TPM2_HMAC;
+                else if (host_key)
+                        id = CRED_AES256_GCM_BY_HOST;
+                else
+                        return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                               "TPM2 not available and host key located on temporary file system, no encryption key available.");
+        } else
+                id = with_key;
+
+        /* Let's now take the host key and the TPM2 key and hash it together, to use as encryption key for the data */
+        r = sha256_hash_host_and_tpm2_key(host_key, host_key_size, tpm2_key, tpm2_key_size, md);
+        if (r < 0)
+                return r;
+
+        assert_se(cc = EVP_aes_256_gcm());
+
+        ks = EVP_CIPHER_key_length(cc);
+        assert(ks == sizeof(md));
+
+        bs = EVP_CIPHER_block_size(cc);
+        assert(bs > 0);
+        assert((size_t) bs <= CREDENTIAL_FIELD_SIZE_MAX);
+
+        ivs = EVP_CIPHER_iv_length(cc);
+        if (ivs > 0) {
+                assert((size_t) ivs <= CREDENTIAL_FIELD_SIZE_MAX);
+
+                iv = malloc(ivs);
+                if (!iv)
+                        return log_oom();
+
+                r = genuine_random_bytes(iv, ivs, RANDOM_BLOCK);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to acquired randomized IV: %m");
+        }
+
+        ts = 16; /* FIXME: On OpenSSL 3 there is EVP_CIPHER_CTX_get_tag_length(), until then let's hardcode this */
```

Comment:
Is the API already set in stone, given it's at beta stage? If so it would be nice to just do the meson-ry to use the function if it is available immediately, as these TODOs tend to be forgotten about

Reply:
Well, i would have,  but i can't really test this yet, as long as fedora doesn't prep an RPM for this yet...

Not sure how set in stone the api is. I noticed that some versions had the the "get" in the name and others without, so i wouldn't wanna bet on it.

Reply:
makes sense